### ISSUE TITLE
'Can't find preLaunchTask' error for first launch item when multiple launch items and tasks are present in JSON configuration

### ISSUE THREAD

**Issue by szul** [microsoft/vscode#94381]
- VSCode Version: 1.43.2
- OS Version: Windows_NT x64 10.0.17134

Version: 1.43.2 (system setup)
Commit: 0ba0ca52957102ca3527cf479571617f0de6ed50
Date: 2020-03-24T07:38:38.248Z
Electron: 7.1.11
Chrome: 78.0.3904.130
Node.js: 12.8.1
V8: 7.8.279.23-electron.0
OS: Windows_NT x64 10.0.17134

Steps to Reproduce:

1. [Launch file](https://gist.github.com/szul/77fe124d00a7e25bbee5574e7f1a27ae)
2. [Task file](https://gist.github.com/szul/024c5a64505dc3f6d24efbdcac9cbb4a)
3. Launch the first item in the `launch.json` file.

Does this issue occur when all extensions are disabled?: [Yes]/No

With the above launch and task configurations, launching the 'Launch OTS Program' runs the `preLaunchTask` and launches fine, but trying to run the 'Launch LMS Program' launch item says it can't fine the `preLaunchTask`. If I reverse the order of the tasks in the `tasks.json` file, it works for the `Launch LMS Program'.

Is there an issue with multiple launch and task items in those configuration files? Seems like it's only running the last task and not the labeled task.

**Comment by alexr00** [comment]
@szul this is because both of your tasks are the same npm script, which makes them the same task regardless of the labels you give them. To have two tasks like this, you'll need to have two different npm scripts in your package.json.

**Comment by szul** [comment]
@alexr00 This is a monorepo with the `.vscode` folder in the top level (root) of the directory. Then I have an `apps` folder with each of the applications in the monorepo. Each of those apps have their own `package.json` file. Notice the `cwd` in the `task.json`. I'm trying to run the `npm build` in the desired directory&mdash;so yes, the "build" might be the same name for script, but those are in different `package.json` files.

Basically, I have a `launch.json` and a `tasks.json` in the top level `.vscode` folder of a monorepo, and I have set up launch configurations and pre-launch tasks to build that individual app and launch that particular application. However, I'm getting the behavior mentioned. It seems like Visual Studio Code is matching on the build task name without understanding that it needs to `cd` into the sub-directory first?

**Comment by szul** [comment]
@alexr00 But to your point, as a temporary fix, if I create an npm script item with a unique name in the sub-folder `package.json` and just have that run the `npm run build` script, it works, but basically I have to do the following:

In each sub-project inside of the monorepository, I have to create another npm script item with a unique build name (e.g., `build-lms`) that simply runs `npm run build`. Then use this unique name as the `preLaunchTask` in the `launch.json` file that is in the root of the monorepository.

So in this scenario, the launch process executes a pre-launch task with a unique name that just calls the build task. It seems like Visual Studio Code doesn't recognize a monorepository with multiple `package.json` files in sub-folders as being unique.

**Comment by alexr00** [comment]
npm scripts with the same name in different package.json files should work.

**Comment by alexr00** [comment]
If you give all the tasks in your different folders the exact same label in your tasks.json, then yes this is a problem. However, by default npm tasks have folder specific label, usually of the format "npm: compile - myFolder". You can see what your npm tasks end up getting labeled as by executing the Run Task command and looking at the labels there. Does that make sense?

**Comment by szul** [comment]
@alexr00 Labels in `tasks.json` are different, the script name that gets executed are the same (but they are different `package.json` files in different sub-folders.

When I use Run Task from the command palette and review the labels, they are indeed different. I try to use them in the `preLaunchTask` and see what happens.

**Comment by szul** [comment]
Sorry, give me this week to test @alexr00's suggestion. (Have to rewrite my launch.json), and I should be able to close this out.

**Comment by EricJizbaMSFT** [comment]
> this is because both of your tasks are the same npm script, which makes them the same task regardless of the labels you give them.

@alexr00 Any particular reason the label is ignored when determining if a task is unique? We had a [similar issue](https://github.com/microsoft/vscode-azurefunctions/issues/2310) filed on the Azure Functions extension - seems like there are quite a few possible scenarios where the task definition is the same, but the "options" property is different.

In addition, "shell" tasks already seem to support this scenario. For example, if I have the below tasks.json, both shell tasks show up during "Run Task", but only one npm task shows up:
<img width="221" alt="Screen Shot 2020-08-26 at 5 06 38 PM" src="https://user-images.githubusercontent.com/11282622/91368690-9c482e00-e7be-11ea-8a73-94049318b0bd.png">

<details>
<summary>Example tasks.json</summary>

```json
{
	"version": "2.0.0",
	"tasks": [
		{
			"type": "npm",
			"script": "install",
			"label": "npm: install",
			"problemMatcher": []
		},
		{
			"type": "npm",
			"script": "install",
			"label": "npm: install (verbose)",
			"problemMatcher": [],
			"options": {
				"env": {
					"npm_config_loglevel": "verbose"
				}
			}
		},
		{
			"label": "Echo 1",
			"type": "shell",
			"command": "echo $NUMBER",
			"problemMatcher": [],
			"options": {
				"env": {
					"NUMBER": "1"
				}
			}
		},
		{
			"label": "Echo 2",
			"type": "shell",
			"command": "echo $NUMBER",
			"problemMatcher": [],
			"options": {
				"env": {
					"NUMBER": "2"
				}
			}
		}
	]
}
```

</details>

**Comment by szul** [comment]
@alexr00 For the purposes of my original issue, setting the names according to your response solves this, so this can be considered closed. 

However, I do share @EricJizbaMSFTso curiosity and concern over why this is implemented in this way.

**Comment by alexr00** [comment]
The only reason I can give for why it works this way is that internally the ID of an extension provided task is composed of the properties listed in the task definition (`script` for npm for example). The label isn't part of the definition, so it isn't part of the ID. I'm not saying this is intuitive, or even the best way to do this, but changing it would likely be destabilizing. Unless I get a lot of feedback from many users saying they want to have the same task with different labels, I wouldn't make the change.
